### PR TITLE
ToolsPanel: Try splitting out ToolsPanelDropdownMenu

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Splits `ToolsPanelDropdownMenu` out from `ToolsPanel`.

```jsx
<ToolsPanel resetAll={ resetAll }>
	<HStack style={{ gridColumn: 'span 2' }}>
		<Button icon={ arrowLeft } label="Back" isSmall />
		<Heading level={ 2 }>Typography</Heading>
		<ToolsPanelDropdownMenu label="Typography settings" />
	</HStack>
	<ToolsPanelItem>...</ToolsPanelItem>
	<ToolsPanelItem>...</ToolsPanelItem>
</ToolsPanel>
```

## Why?
So that consumers can implement their own header design while using ToolsPanel. See https://github.com/WordPress/gutenberg/pull/44180 for one such example of when we want to do this.

## How?
- `ToolsPanelHeader` is split up into `ToolsPanelHeader` and `ToolsPanelDropdownMenu`.
- The `resetAll` and `toggleItem` props are removed from `ToolsPanelHeader`. These are now accessed via `ToolsPanelContext`.
- We should probably mark the `label` prop on `ToolsPanel` as optional.
- I'm unsure if we should go even further and remove the `Grid` from `ToolsPanel`.
  - Pro: Further separation of "how the panel looks" from "tool panel toggle logic".
  - Con: `ToolsPanel` wouldn't even really be a panel anymore?
  - Con: Difficult to untangle some of the logic in `ToolsPanel`. In particular it uses `menuItems` in order to calculate the `className`.

## Testing Instructions
```
npm run storybook:dev
```

## Screenshots or screencast 
Screen Shot 2022-09-20 at 15 43 34

### CODE REVIEW

**Comment by noisysocks** [comment]
@ciampo: Here's a rough attempt at implementing what was discussed in https://github.com/WordPress/gutenberg/pull/44180. What do you think?

I tried to go further and split out the `Grid` from `ToolsPanel` so that we have both `ToolsPanel` (provides a grid) and `ToolsPanelProvider` (no grid) but had difficulty separating some of the `className` logic that `ToolsPanel` has.

I'm not 100% sure about the direction. It works and is composition-ey, but, idk... does it make `ToolsPanel` too unopinionated? A `ToolsPanel` should provide `Panel` bits, no?

Would a simple prop be better?

```jsx
<ToolsPanel
	heading={ <>
		<Button icon={ arrowLeft } label="Back" isSmall />
		<Heading level={ 2 }>Typography</Heading>		
	</> }
	resetAll={ resetAll }
>
	<ToolsPanelItem>...</ToolsPanelItem>
	<ToolsPanelItem>...</ToolsPanelItem>
</ToolsPanel>
```

Or, maybe I could use absolute positioning to fake the effect that I'm after in https://github.com/WordPress/gutenberg/pull/44180? 😅

File: packages/components/src/tools-panel/tools-panel-dropdown-menu/hook.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import * as styles from '../styles';
+import { useToolsPanelContext } from '../context';
+import { useContextSystem, WordPressComponentProps } from '../../ui/context';
+import { useCx } from '../../utils/hooks/use-cx';
+import type { ToolsPanelDropdownMenuProps } from '../types';
+
+export function useToolsPanelDropdownMenu(
+	props: WordPressComponentProps< ToolsPanelDropdownMenuProps, 'div', false >
+) {
+	const { ...otherProps } = useContextSystem(
```
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel-dropdown-menu/hook.ts:18]: Nit: we don't need the rest operator here

File: packages/components/src/tools-panel/types.ts [hunk 2]
```diff
@@ -49,26 +49,16 @@ export type ToolsPanelProps = {
...
-	resetAll: ResetAll;
-	/**
-	 * This is executed when an individual control's menu item is toggled. It
-	 * will update the panel's menu item state and call the panel item's
-	 * `onSelect` or `onDeselect` callbacks as appropriate.
-	 */
-	toggleItem: ( label: string ) => void;
 };
 
+export type ToolsPanelHeaderProps = ToolsPanelDropdownMenuProps;
```
Comment (ciampo) [inline packages/components/src/tools-panel/types.ts:60]: Nit: we'd probably need to duplicate the type here, since we'd want the description for the `label` prop to be different for the two components

**Review by ciampo — COMMENTED** [review]
Thank you for trying this out! Code changes are nice and clean, and easy to follow thanks to the detailed PR description

> I tried to go further and split out the `Grid` from `ToolsPanel` so that we have both `ToolsPanel` (provides a grid) and `ToolsPanelProvider` (no grid) but had difficulty separating some of the `className` logic that `ToolsPanel` has.
> 
> I'm not 100% sure about the direction. It works and is composition-ey, but, idk... does it make `ToolsPanel` too unopinionated? A `ToolsPanel` should provide `Panel` bits, no?

I'm not fully convinced either. Let's keep the `Grid` wrapper as part of `ToolsPanel` for now, we can always come back to this at a later point.

> Would a simple prop be better?

It's definitely a possibility, although I'd personally prefer using `children` instead of render props, as it feels cleaner to me.

Seeing the route that we're taking here, I think that we should just stop using `ToolsPanelHeader` internally, and always expect consumers of `ToolsPanel` to define their own label explicitly (this would also mean removing the `label` prop from `ToolsPanel`).

In order to avoid too much disruption during the "transition", we could:
 1. on this PR:
    - keep the newly created `ToolsPanelDropdownMenu` as-is on this PR
    - refactor `ToolsPanelHeader` to behave like currently on `trunk`, but refactoring it to use `ToolsPanelDropdownMenu` internally
    - refactor most Storybook examples and tests to avoid using the `label` prop
    - this way, current usages of `ToolsPanel` would keep working as expected
 2. Refactor usages of `ToolsPanel` across the codebase to stop using the `label` prop, and define their own header as part of `children` instead (including using `ToolsPanelDropdownMenu`)
 3. Formally deprecate the `label` prop (outputting a console warning) and scheduling the prop for deletion in 2 WordPress versions time.
 4. When removing the `label` prop, we'll be also able to remove the `ToolsPanelHeader` component altogether.

---

What do you think? Tagging @aaronrobertshaw and @mirka for thoughts too.

**Comment by noisysocks** [comment]
Thanks for your thoughts! API design feels more like art than science so extra opinions are really helpful.

> I'm not fully convinced either. Let's keep the `Grid` wrapper as part of `ToolsPanel` for now, we can always come back to this at a later point.

👍 sounds like a plan.

> always expect consumers of `ToolsPanel` to define their own label explicitly (this would also mean removing the `label` prop from `ToolsPanel`).

My worry with this is that we're making `ToolsPanel` too difficult to use for the majority use-case which is that you have a simple text header. Maybe we should also provide `ToolsPanelHeading` (a `Heading` with the right font size) and `ToolsPanelRow` (a `View` with `grid-column: span 2`) to make things a little bit more frictionless for consumers.

In other words this:

```jsx
<ToolsPanel resetAll={ resetAll }>
	<HStack style={{ gridColumn: 'span 2' }}>
		<Heading level={ 2 } style={{ fontSize: 'inherit', fontWeight: 500, lineHeight: 'normal' }}>
			Typography
		</Heading>
		<ToolsPanelDropdownMenu label="Typography settings" />
	</HStack>
	<ToolsPanelItem>...</ToolsPanelItem>
	<ToolsPanelItem>...</ToolsPanelItem>
</ToolsPanel>
```

Becomes this:

```jsx
<ToolsPanel resetAll={ resetAll }>
	<ToolsPanelRow>
		<HStack>
			<ToolsPanelHeading>Typography</ToolsPanelHeading>
			<ToolsPanelDropdownMenu label="Typography settings" />
		</HStack>
	</ToolsPanelRow>
	<ToolsPanelItem>...</ToolsPanelItem>
	<ToolsPanelItem>...</ToolsPanelItem>
</ToolsPanel>
```

**Comment by noisysocks** [comment]
After some more exploration I realised this won't work for https://github.com/WordPress/gutenberg/pull/44180 because what I actually want is for the header to appear _completely_ outside of the `ToolsPanel` because `ToolsPanel` provides its own paddings and margins and whatnot which doesn't match the rest of Global Styles 😅

```html
<HStack>
	<NavigatorBackButton />
	<Heading />
	<ToolsPanelDropdownMenu />
</HStack>
<ToolsPanel>
	<ToolsPanelItem />
	<ToolsPanelItem />
</ToolsPanel>
```

I think the only solution to combining the headers in https://github.com/WordPress/gutenberg/pull/44180 is to have it so that a consumer can use the `ToolsPanel` logic without using the `ToolsPanel` UI. For example the Typography panel could call `useToolsPanel`, feed `toolsPanelContext` into `ToolsPanelContext.Provider`, and then roll its own `Grid`, header, etc.

To be honest though this is becoming more work than I think it's worth so I'll put this on ice and explore some other options first 🙂

**Comment by ciampo** [comment]
> My worry with this is that we're making ToolsPanel too difficult to use for the majority use-case which is that you have a simple text header. Maybe we should also provide ToolsPanelHeading (a Heading with the right font size) and ToolsPanelRow (a View with grid-column: span 2) to make things a little bit more frictionless for consumers.

Yes, that would be a good idea. Basically, `ToolsPanel` would offer all the ready-made subcomponents needed for implementing the "default" usecase. But any consumer would be able to swap these ready-made components for their own custom versions if they wanted so.

> what I actually want is for the header to appear _completely_ outside of the `ToolsPanel` because `ToolsPanel` provides its own paddings and margins and whatnot which doesn't match the rest of Global Styles 😅

Yeah, the only way there would be to completely destructure `ToolsPanel` and allow consumers to use its parts across a wider UI (as you suggested)

> To be honest though this is becoming more work than I think it's worth so I'll put this on ice and explore some other options first 🙂

Sure — feel free to ping again if/when you resume work on this, always happy to help

**Comment by noisysocks** [comment]
No such thing as a failed experiment 😀 thanks for the thoughts @ciampo

**Comment by ciampo** [comment]
> No such thing as a failed experiment 😀 thanks for the thoughts @ciampo

Nonetheless a very interesting API thinking exercise, it may be useful in the future — who knows!

Thank you 🙏
